command/clistate/state.go

```diff
@@ -48,47 +50,119 @@ that no one else is holding a lock.
 `
 )
 
-// Lock locks the given state and outputs to the user if locking
-// is taking longer than the threshold.  The lock is retried until the context
-// is cancelled.
-func Lock(ctx context.Context, s state.State, info *state.LockInfo, ui cli.Ui, color *colorstring.Colorize) (string, error) {
-	var lockID string
+// Locker allows for more convenient locking, by creating the timeout context
+// and LockInfo for the caller, while storing any related data required for
+// Unlock.
+type Locker interface {
+	// Lock the provided state, storing the reason string in the LockInfo.
```

Comment:
This comment was a little confusing to me at first because "the LockInfo" seems like an implementation detail of this function, rather than something visible on this interface. However, reading the code below I see that it gets created and passed into the `LockWithContext` function; I think it could help to clarify a little what side-effects this function has for the given state, since the interactions between subsystems here aren't obvious.

Reply:
Ah, I see the confusion. It's because it _used_ to be visible in the interface -- you have to pass a `LockInfo` to the state `Lock` method. I'll clarify this a little more